### PR TITLE
[SNOW-851840] Enable tombstone record ingestion in Snowpipe Streaming

### PR DESCRIPTION
**Pull request by sfc-gh-rcheng**
Allows tombstone records (records with NULL value and value schema) to be ingested.

- Main change in RecordService creates an empty SnowflakeRecordValue when given a null value or schema

Motivation:
- 6 customers across slack requests and github issues 
  - https://github.com/snowflakedb/snowflake-kafka-connector/issues/481
  - https://github.com/snowflakedb/snowflake-kafka-connector/issues/416
  - https://github.com/snowflakedb/snowflake-kafka-connector/issues/599
- [1 pager](https://docs.google.com/document/d/1pvaQi6NaXGxqom2BrUS2oMsvyh9QTdHPGEUHqRo6aS0/edit?usp=sharing)

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java [hunk 1]
```diff
@@ -172,18 +172,6 @@ public void start(final Map<String, String> parsedConfig) {
     final long bufferFlushTime =
         Long.parseLong(parsedConfig.get(SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC));
 
-    // Falling back to default behavior which is to ingest an empty json string if we get null
-    // value. (Tombstone record)
-    SnowflakeSinkConnectorConfig.BehaviorOnNullValues behavior =
-        SnowflakeSinkConnectorConfig.BehaviorOnNullValues.DEFAULT;
-    if (parsedConfig.containsKey(SnowflakeSinkConnectorConfig.BEHAVIOR_ON_NULL_VALUES_CONFIG)) {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:179]: BehaviorOnNullValues (aka ingest or don't ingest tombstone records) is only used in RecordService. Thus we don't need `setBehaviorOnNullValuesConfig` and can just pass the config into RecordService

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 2]
```diff
@@ -924,6 +924,19 @@ public static String getExceptionMessage(String customMessage, Exception ex) {
...
   }
 
+  public static SnowflakeSinkConnectorConfig.BehaviorOnNullValues getBehaviorOnNullValuesEnum(
+      Map<String, String> config) {
+    return Arrays.stream(SnowflakeSinkConnectorConfig.BehaviorOnNullValues.values())
+        .filter(
+            behavior ->
+                behavior
+                    .toString()
+                    .equalsIgnoreCase(
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:934]: needed because we need to convert the config param to the enum with an `equalsIgnoreCase`.

Alternatively I can do something like the following since we only have two options in the enum, however it's not optimal if we decide to add any more behaviors

```
return config.get(BEHAVIOR_ON_NULL_VALUES_CONFIG).equalsIgnoreCase(SnowflakeSinkConnectorConfig.BehaviorOnNullValues.IGNORE.toString()) ?
        SnowflakeSinkConnectorConfig.BehaviorOnNullValues.IGNORE :
        SnowflakeSinkConnectorConfig.BehaviorOnNullValues.DEFAULT;
```
return

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 3]
```diff
@@ -89,60 +86,47 @@ public class RecordService {
...
 
   /**
    * process records output JSON format: { "meta": { "offset": 123, "topic": "topic name",
    * "partition": 123, "key":"key name" } "content": "record content" }
    *
    * <p>create a JsonRecordService instance
-   *
-   * @param telemetryService Telemetry Service Instance. Can be null.
    */
-  public RecordService(SnowflakeTelemetryService telemetryService) {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:104]: telemetryService is never used in RecordService, so I removed it

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 4]
```diff
@@ -152,19 +136,25 @@ public void setEnableSchematization(final boolean enableSchematization) {
...
     if (record.value() == null || record.valueSchema() == null) {
-      throw SnowflakeErrors.ERROR_5016.getException();
-    }
-    if (!record.valueSchema().name().equals(SnowflakeJsonSchema.NAME)) {
-      throw SnowflakeErrors.ERROR_0009.getException();
-    }
-    if (!(record.value() instanceof SnowflakeRecordContent)) {
-      throw SnowflakeErrors.ERROR_0010.getException(
-          "Input record should be SnowflakeRecordContent object");
+      if (this.ingestTombstoneRecords) {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:142]: This is the crux of the PR - if we want to ingestTombstoneRecords, create an empty record

File: src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java [hunk 5]
```diff
@@ -146,173 +148,6 @@ public void testIngestion() throws Exception {
     //    assert !conn.pipeExist(pipe);
   }
 
-  @Test
-  public void testTombstoneRecords_DEFAULT_behavior_ingestion_SFJsonConverter() throws Exception {
```
Comment (sfc-gh-rcheng) [inline src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java:150]: moved to TombstoneRecordIngestionIT.java

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 6]
```diff
@@ -108,13 +109,12 @@ public RecordService(SnowflakeTelemetryService telemetryService) {
   /** Record service with null telemetry Service, only use it for testing. */
   @VisibleForTesting
   public RecordService() {
-    this.telemetryService = null;
+    this(null);
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:112]: set default here?

Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:112]: I don't think that is necessary as it its set to default on line 75?

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 7]
```diff
@@ -89,7 +91,6 @@ public class RecordService {
 
   // This class is designed to work with empty metadata config map
   private SnowflakeMetadataConfig metadataConfig = new SnowflakeMetadataConfig();
-
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:92]: Is this change intended?

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 8]
```diff
@@ -108,13 +109,12 @@ public RecordService(SnowflakeTelemetryService telemetryService) {
...
   @VisibleForTesting
   public RecordService() {
-    this.telemetryService = null;
+    this(null);
   }
 
   public void setMetadataConfig(SnowflakeMetadataConfig metadataConfigIn) {
     metadataConfig = metadataConfigIn;
   }
-
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:117]: same here

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 9]
```diff
@@ -297,7 +316,7 @@ public SnowflakeTableRow(SnowflakeRecordContent content, JsonNode metadata) {
   }
 
   void putKey(SinkRecord record, ObjectNode meta) {
-    if (record.key() == null) {
+    if (record.key() == null || record.keySchema() == null) {
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:319]: why we can't put the key when the keySchema is null but not the key? This change might be risky and introduce a behavior change for both Snowpipe and Snowpipe Streaming

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM, only comment is about a potential risky change which I'm not sure if it's necessary

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java (hunk 7 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java]: reverted, thanks for the catch

File: src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java [hunk 10]
```diff
@@ -749,19 +579,23 @@ public void testNativeNullIngestion() throws Exception {
     service.closeAll();
   }
 
-  @Test(expected = SnowflakeKafkaConnectorException.class)
```
Comment (sfc-gh-rcheng) [inline src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java:752]: We need to allow null value ingestion because community converters create null values instead of empty SnowflakeRecordContents. Currently Snowpipe fails to ingest null value records with community converters.

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java (hunk 9 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java]: a null keySchema will NPE on the next if statement, however we should never get into this situation due to previous handleNativeRecord() calls. I'll update the logic here to both put the key and handle the NPE

Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java]: the current code will NPE on the next if statement if keySchema is null. 

handleNativeRecord() prevents null keySchemas from being passed into putKey, however we should throw a better error if we ever get a null keySchema, so I added a null check.

This shouldn't be risky because we are changing the error thrown from NPE to a Snowflake error.

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java [hunk 11]
```diff
@@ -301,6 +322,12 @@ void putKey(SinkRecord record, ObjectNode meta) {
       return;
     }
 
+    if (record.keySchema() == null) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: why do we need this?

File: src/test/java/com/snowflake/kafka/connector/ConnectorConfigTest.java [hunk 12]
```diff
@@ -16,11 +16,59 @@
...
+      this.name = name;
+      this.converter = converter;
+    }
+
+    public String toString() {
+      return this.name;
+    }
+  };
+
+  public enum CustomSfConverter {
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/ConnectorConfigTest.java:48]: This is fine but we do have
` public static final Set<String> CUSTOM_SNOWFLAKE_CONVERTERS ` which can be resued.

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java (hunk 11 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: I am concerned about this change.

Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: discussed in this comment - ensuring we throw a snowflake error instead of NPE. We should never get into this situation, but i want to add this just in case

File: src/test/java/com/snowflake/kafka/connector/ConnectorConfigTest.java (hunk 12 above)
Comment (sfc-gh-rcheng) [reply src/test/java/com/snowflake/kafka/connector/ConnectorConfigTest.java:48]: i considered using that, but i preferred having the enum

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java (hunk 11 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: keyschemas are often null
I dont think we should do this. this seems like a behavior change.

Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: keyschemas are often null, we update it in handleNativeRecords earlier in the call stack. if the key schema is null in this method the next if statement throws an NPE. 

The only behavior change here is to change the exception thrown from an NPE to a Snowflake Error

Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: I See, thanks. Sounds like a harmless change but its recommended to not have refactoring/fixes in the same PR as features. 

This is a small enough change to go through but it just raises more questions from reviewers POV

**Review by sfc-gh-japatel — APPROVED** [review]
lgtm

File: src/main/java/com/snowflake/kafka/connector/records/RecordService.java (hunk 11 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/records/RecordService.java:325]: I see, it was caught during testing that method since i wanted to give all possible null scenarios

**Comment by sfc-gh-rcheng** [comment]
@aidan-melen this should go out in release 2.0.2, which should be released in the next couple weeks
